File: internal/tools/calendar/calendar.go
package calendar

import (
	"encoding/json"
	"fmt"
	"log"
	"os"
	"path/filepath"
	"strings"
	"time"
)

type Schedule struct {
	ID          string    `json:"id"`
	UserID      string    `json:"user_id"`
	Title       string    `json:"title"`
	Description string    `json:"description"`
	StartTime   time.Time `json:"start_time"`
	EndTime     time.Time `json:"end_time"`
	Tags        []string  `json:"tags"`
}

type CalendarManager struct {
	dataPath  string
	schedules []Schedule
}

func NewCalendarManager() (*CalendarManager, error) {
	workingDir, err := os.Getwd()
	if err != nil {
		log.Printf("Error getting working directory: %v\n", err)
		return nil, err
	}

	dataDir := filepath.Join(workingDir, "data", "calendar")
	if err := os.MkdirAll(dataDir, 0755); err != nil {
		log.Printf("Error creating calendar directory: %v\n", err)
		return nil, err
	}

	dataPath := filepath.Join(dataDir, "calendar.json")
	manager := &CalendarManager{
		dataPath:  dataPath,
		schedules: []Schedule{},
	}

	if err := manager.loadSchedules(); err != nil {
		log.Printf("Warning: error loading schedules: %v", err)
	}

	return manager, nil
}

func (cm *CalendarManager) loadSchedules() error {
	if _, err := os.Stat(cm.dataPath); err == nil {
		data, err := os.ReadFile(cm.dataPath)
		if err != nil {
			return fmt.Errorf("error reading data file: %v", err)
		}

		if len(data) == 0 {
			return nil
		}

		if err := json.Unmarshal(data, &cm.schedules); err != nil {
			log.Printf("Warning: error parsing data file: %v, returning empty data", err)
			return nil
		}
	}
	return nil
}

func (cm *CalendarManager) saveSchedules() error {
	if cm.schedules == nil {
		cm.schedules = make([]Schedule, 0)
	}

	data, err := json.MarshalIndent(cm.schedules, "", "  ")
	if err != nil {
		return fmt.Errorf("error marshaling data: %v", err)
	}

	dir := filepath.Dir(cm.dataPath)
	if err := os.MkdirAll(dir, 0755); err != nil {
		return fmt.Errorf("error creating directory: %v", err)
	}

	if err := os.WriteFile(cm.dataPath, data, 0644); err != nil {
		return fmt.Errorf("error writing data file: %v", err)
	}

	return nil
}

func (cm *CalendarManager) AddSchedule(schedule Schedule) error {
	cm.schedules = append(cm.schedules, schedule)
	return cm.saveSchedules()
}

func (cm *CalendarManager) UpdateSchedule(id string, userID string, updatedSchedule Schedule) error {
	for i, s := range cm.schedules {
		if s.ID == id && s.UserID == userID {
			cm.schedules[i] = updatedSchedule
			return cm.saveSchedules()
		}
	}
	return fmt.Errorf("schedule with ID %s not found", id)
}

func (cm *CalendarManager) DeleteSchedule(id string, userID string) error {
	for i, s := range cm.schedules {
		if s.ID == id && s.UserID == userID {
			cm.schedules = append(cm.schedules[:i], cm.schedules[i+1:]...)
			return cm.saveSchedules()
		}
	}
	return fmt.Errorf("schedule with ID %s not found", id)
}

func (cm *CalendarManager) SearchByDateRange(userID string, start, end time.Time) []Schedule {
	var results []Schedule
	for _, s := range cm.schedules {
		if s.UserID == userID && (s.StartTime.Equal(start) || s.StartTime.After(start)) && (s.EndTime.Equal(end) || s.EndTime.Before(end)) {
			results = append(results, s)
		}
	}
	return results
}

func (cm *CalendarManager) SearchByTitle(userID string, title string) []Schedule {
	var results []Schedule
	searchTitle := strings.ToLower(title)
	for _, s := range cm.schedules {
		if s.UserID == userID && strings.Contains(strings.ToLower(s.Title), searchTitle) {
			results = append(results, s)
		}
	}
	return results
}

func (cm *CalendarManager) SearchByTags(userID string, tags []string) []Schedule {
	var results []Schedule
	for _, s := range cm.schedules {
		if s.UserID != userID {
			continue
		}
		for _, tag := range tags {
			for _, scheduleTag := range s.Tags {
				if tag == scheduleTag {
					results = append(results, s)
					break
				}
			}
		}
	}
	return results
}

type CalendarTool struct {
	manager *CalendarManager
}

func NewCalendarTool() *CalendarTool {
	manager, err := NewCalendarManager()
	if err != nil {
		log.Printf("Error creating calendar manager: %v", err)
		return nil
	}
	return &CalendarTool{manager: manager}
}

func (ct *CalendarTool) CallTool(arguments string) string {
	var params map[string]interface{}
	if err := json.Unmarshal([]byte(arguments), &params); err != nil {
		return fmt.Sprintf("Error parsing arguments: %v", err)
	}

	action, ok := params["action"].(string)
	if !ok {
		return "Error: action not found"
	}

	switch action {
	case "add_schedule":
		return ct.handleAddSchedule(params)
	case "update_schedule":
		return ct.handleUpdateSchedule(params)
	case "delete_schedule":
		return ct.handleDeleteSchedule(params)
	case "search_by_date":
		return ct.handleSearchByDate(params)
	case "search_by_title":
		return ct.handleSearchByTitle(params)
	case "search_by_tags":
		return ct.handleSearchByTags(params)
	default:
		return fmt.Sprintf("Error: invalid action: %s", action)
	}
}

func (ct *CalendarTool) handleAddSchedule(params map[string]interface{}) string {
	userID, ok := params["user_id"].(string)
	if !ok || userID == "" {
		return "Error: user_id is required"
	}

	scheduleData, ok := params["schedule"].(map[string]interface{})
	if !ok {
		return "Error: invalid schedule data"
	}

	title, ok := scheduleData["title"].(string)
	if !ok {
		return "Error: title is required"
	}

	description, ok := scheduleData["description"].(string)
	if !ok {
		return "Error: description is required"
	}

	startTimeStr, ok := scheduleData["start_time"].(string)
	if !ok {
		return "Error: start_time is required"
	}
	startTime, err := time.Parse(time.RFC3339, startTimeStr)
	if err != nil {
		return fmt.Sprintf("Error: invalid start_time format: %v", err)
	}

	endTimeStr, ok := scheduleData["end_time"].(string)
	if !ok {
		return "Error: end_time is required"
	}
	endTime, err := time.Parse(time.RFC3339, endTimeStr)
	if err != nil {
		return fmt.Sprintf("Error: invalid end_time format: %v", err)
	}

	tagsInterface, ok := scheduleData["tags"].([]interface{})
	if !ok {
		return "Error: tags is required"
	}
	tags := make([]string, len(tagsInterface))
	for i, tag := range tagsInterface {
		tags[i], ok = tag.(string)
		if !ok {
			return "Error: invalid tag format"
		}
	}

	schedule := Schedule{
		ID:          fmt.Sprintf("%d", time.Now().UnixNano()),
		UserID:      userID,
		Title:       title,
		Description: description,
		StartTime:   startTime,
		EndTime:     endTime,
		Tags:        tags,
	}

	if err := ct.manager.AddSchedule(schedule); err != nil {
		return fmt.Sprintf("Error adding schedule: %v", err)
	}

	return "Schedule added successfully"
}

func (ct *CalendarTool) handleUpdateSchedule(params map[string]interface{}) string {
	userID, ok := params["user_id"].(string)
	if !ok || userID == "" {
		return "Error: user_id is required"
	}

	scheduleData, ok := params["schedule"].(map[string]interface{})
	if !ok {
		return "Error: invalid schedule data"
	}

	id, ok := scheduleData["id"].(string)
	if !ok {
		return "Error: id is required"
	}

	title, ok := scheduleData["title"].(string)
	if !ok {
		return "Error: title is required"
	}

	description, ok := scheduleData["description"].(string)
	if !ok {
		return "Error: description is required"
	}

	startTimeStr, ok := scheduleData["start_time"].(string)
	if !ok {
		return "Error: start_time is required"
	}
	startTime, err := time.Parse(time.RFC3339, startTimeStr)
	if err != nil {
		return fmt.Sprintf("Error: invalid start_time format: %v", err)
	}

	endTimeStr, ok := scheduleData["end_time"].(string)
	if !ok {
		return "Error: end_time is required"
	}
	endTime, err := time.Parse(time.RFC3339, endTimeStr)
	if err != nil {
		return fmt.Sprintf("Error: invalid end_time format: %v", err)
	}

	tagsInterface, ok := scheduleData["tags"].([]interface{})
	if !ok {
		return "Error: tags is required"
	}
	tags := make([]string, len(tagsInterface))
	for i, tag := range tagsInterface {
		tags[i], ok = tag.(string)
		if !ok {
			return "Error: invalid tag format"
		}
	}

	schedule := Schedule{
		ID:          id,
		UserID:      userID,
		Title:       title,
		Description: description,
		StartTime:   startTime,
		EndTime:     endTime,
		Tags:        tags,
	}

	if err := ct.manager.UpdateSchedule(id, userID, schedule); err != nil {
		return fmt.Sprintf("Error updating schedule: %v", err)
	}

	return "Schedule updated successfully"
}

func (ct *CalendarTool) handleDeleteSchedule(params map[string]interface{}) string {
	userID, ok := params["user_id"].(string)
	if !ok || userID == "" {
		return "Error: user_id is required"
	}

	id, ok := params["schedule_id"].(string)
	if !ok {
		return "Error: schedule_id is required"
	}

	if err := ct.manager.DeleteSchedule(id, userID); err != nil {
		return fmt.Sprintf("Error deleting schedule: %v", err)
	}

	return "Schedule deleted successfully"
}

func (ct *CalendarTool) handleSearchByDate(params map[string]interface{}) string {
	userID, ok := params["user_id"].(string)
	if !ok || userID == "" {
		return "Error: user_id is required"
	}

	dateRange, ok := params["date_range"].(map[string]interface{})
	if !ok {
		return "Error: date_range is required"
	}

	startStr, ok := dateRange["start"].(string)
	if !ok {
		return "Error: start date is required"
	}
	start, err := time.Parse(time.RFC3339, startStr)
	if err != nil {
		return fmt.Sprintf("Error: invalid start date format: %v", err)
	}

	endStr, ok := dateRange["end"].(string)
	if !ok {
		return "Error: end date is required"
	}
	end, err := time.Parse(time.RFC3339, endStr)
	if err != nil {
		return fmt.Sprintf("Error: invalid end date format: %v", err)
	}

	schedules := ct.manager.SearchByDateRange(userID, start, end)
	result, err := json.Marshal(schedules)
	if err != nil {
		return fmt.Sprintf("Error marshaling results: %v", err)
	}

	return string(result)
}

func (ct *CalendarTool) handleSearchByTitle(params map[string]interface{}) string {
	userID, ok := params["user_id"].(string)
	if !ok || userID == "" {
		return "Error: user_id is required"
	}

	title, ok := params["title"].(string)
	if !ok {
		return "Error: title is required"
	}

	schedules := ct.manager.SearchByTitle(userID, title)
	result, err := json.Marshal(schedules)
	if err != nil {
		return fmt.Sprintf("Error marshaling results: %v", err)
	}

	return string(result)
}

func (ct *CalendarTool) handleSearchByTags(params map[string]interface{}) string {
	userID, ok := params["user_id"].(string)
	if !ok || userID == "" {
		return "Error: user_id is required"
	}

	tagsInterface, ok := params["tags"].([]interface{})
	if !ok {
		return "Error: tags is required"
	}

	tags := make([]string, len(tagsInterface))
	for i, tag := range tagsInterface {
		tags[i], ok = tag.(string)
		if !ok {
			return "Error: invalid tag format"
		}
	}

	schedules := ct.manager.SearchByTags(userID, tags)
	result, err := json.Marshal(schedules)
	if err != nil {
		return fmt.Sprintf("Error marshaling results: %v", err)
	}

	return string(result)
}

File: .teo/skills/calendar/scripts/calendar.py
import sys
import json
import os
import datetime
import time

# Constants
DATA_DIR = os.path.join("data", "calendar")
DATA_FILE = os.path.join(DATA_DIR, "calendar.json")

def ensure_data_dir():
    if not os.path.exists(DATA_DIR):
        os.makedirs(DATA_DIR)

def load_schedules():
    ensure_data_dir()
    if os.path.exists(DATA_FILE):
        try:
            with open(DATA_FILE, 'r') as f:
                data = json.load(f)
                if isinstance(data, list):
                    return data
        except Exception:
            return []
    return []

def save_schedules(schedules):
    ensure_data_dir()
    with open(DATA_FILE, 'w') as f:
        json.dump(schedules, f, indent=2)

def add_schedule(user_id, schedule_data):
    schedules = load_schedules()
    
    # Validation
    required_fields = ["title", "description", "start_time", "end_time", "tags"]
    for field in required_fields:
        if field not in schedule_data:
            return json.dumps({"error": f"Error: {field} is required"})
            
    # Create Schedule object
    new_schedule = {
        "id": str(time.time_ns()),
        "user_id": user_id,
        "title": schedule_data["title"],
        "description": schedule_data["description"],
        "start_time": schedule_data["start_time"],
        "end_time": schedule_data["end_time"],
        "tags": schedule_data.get("tags", [])
    }
    
    schedules.append(new_schedule)
    save_schedules(schedules)
    return json.dumps({"message": "Schedule added successfully"})

def update_schedule(user_id, schedule_data):
    if "id" not in schedule_data:
         return json.dumps({"error": "Error: id is required"})
         
    schedules = load_schedules()
    target_id = schedule_data["id"]
    
    for i, s in enumerate(schedules):
        if s["id"] == target_id and s["user_id"] == user_id:
            # Update fields
            schedules[i]["title"] = schedule_data.get("title", s["title"])
            schedules[i]["description"] = schedule_data.get("description", s["description"])
            schedules[i]["start_time"] = schedule_data.get("start_time", s["start_time"])
            schedules[i]["end_time"] = schedule_data.get("end_time", s["end_time"])
            schedules[i]["tags"] = schedule_data.get("tags", s["tags"])
            
            save_schedules(schedules)
            return json.dumps({"message": "Schedule updated successfully"})
            
    return json.dumps({"error": f"schedule with ID {target_id} not found"})

def delete_schedule(user_id, schedule_id):
    schedules = load_schedules()
    original_len = len(schedules)
    
    schedules = [s for s in schedules if not (s["id"] == schedule_id and s["user_id"] == user_id)]
    
    if len(schedules) == original_len:
         return json.dumps({"error": f"schedule with ID {schedule_id} not found"})
         
    save_schedules(schedules)
    return json.dumps({"message": "Schedule deleted successfully"})

def search_by_date(user_id, start_str, end_str):
    schedules = load_schedules()
    results = []
    
    try:
        # Using string comparison for ISO format dates (RFC3339) is usually safe if consistent
        # But let's try to be a bit more robust if needed, though simple string compare works for ISO8601
        pass
    except Exception as e:
        return json.dumps({"error": f"Date parsing error: {e}"})

    for s in schedules:
        if s["user_id"] == user_id:
            # Check overlap or containment. 
            # Logic from Go: (s.StartTime >= start) && (s.EndTime <= end) ? 
            # Go logic: (s.StartTime.Equal(start) || s.StartTime.After(start)) && (s.EndTime.Equal(end) || s.EndTime.Before(end))
            # Meaning schedule must be entirely within the range? Or start/end within range?
            # Actually Go logic was: s.StartTime >= start AND s.EndTime <= end. So "contained within".
            
            if s["start_time"] >= start_str and s["end_time"] <= end_str:
                results.append(s)
                
    return json.dumps(results, indent=2)

def search_by_title(user_id, title_query):
    schedules = load_schedules()
    results = []
    title_query = title_query.lower()
    
    for s in schedules:
        if s["user_id"] == user_id and title_query in s["title"].lower():
            results.append(s)
            
    return json.dumps(results, indent=2)

def search_by_tags(user_id, tags):
    schedules = load_schedules()
    results = []
    
    for s in schedules:
        if s["user_id"] == user_id:
            for tag in tags:
                if tag in s.get("tags", []):
                    results.append(s)
                    break
    return json.dumps(results, indent=2)

def main():
    if len(sys.argv) < 2:
        print(json.dumps({"error": "No arguments provided"}))
        return

    try:
        args = json.loads(sys.argv[1])
    except json.JSONDecodeError:
        print(json.dumps({"error": "Invalid JSON arguments"}))
        return

    action = args.get("action", "")
    user_id = args.get("user_id", "")
    
    if not user_id:
        print(json.dumps({"error": "Error: user_id is required"}))
        return

    if action == "add_schedule":
        print(add_schedule(user_id, args.get("schedule", {})))
    elif action == "update_schedule":
        print(update_schedule(user_id, args.get("schedule", {})))
    elif action == "delete_schedule":
        print(delete_schedule(user_id, args.get("schedule_id", "")))
    elif action == "search_by_date":
        date_range = args.get("date_range", {})
        print(search_by_date(user_id, date_range.get("start", ""), date_range.get("end", "")))
    elif action == "search_by_title":
        print(search_by_title(user_id, args.get("title", "")))
    elif action == "search_by_tags":
        print(search_by_tags(user_id, args.get("tags", [])))
    else:
        print(json.dumps({"error": f"Error: invalid action: {action}"}))

if __name__ == "__main__":
    main()
